pkg/guacanalytics/patchPlanning.go

```diff
@@ -0,0 +1,162 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package guacanalytics
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Khan/genqlient/graphql"
+	model "github.com/guacsec/guac/pkg/assembler/clients/generated"
+	"github.com/guacsec/guac/pkg/misc/depversion"
+)
+
+type DfsNode struct {
+	expanded bool // true once all node neighbors are added to queue
+	Parent   string
+	depth    int
+	nodeType string // packageName, packageVersion
+}
+
+var path []string
+var nodeMap map[string]DfsNode
+var now string
+var nowNode DfsNode
+var queue []string
+
+// TODO: make more robust usuing predicates
+func SearchDependenciesFromStartNode(ctx context.Context, gqlclient graphql.Client, startID string, stopID string, startType string, maxDepth int) (map[string]DfsNode, error) {
+	startNode, err := model.Node(ctx, gqlclient, startID)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed getting intial node with given ID:%v", err)
+	}
+
+	_, ok := startNode.Node.(*model.NodeNodePackage)
+
+	if !ok {
+		return nil, fmt.Errorf("Not a package")
+	}
+	queue = make([]string, 0) // the queue of nodes in bfs
+
+	nodeMap = map[string]DfsNode{}
+
+	nodeMap[startID] = DfsNode{
+		nodeType: startType,
+	}
+	queue = append(queue, startID)
+
+	for len(queue) > 0 {
+		now = queue[0]
+		queue = queue[1:]
```

Comment:
where else are nodes being added to the frontier here? it seems like this is only a single step?

Reply:
e.g. https://go.dev/tour/methods/4

Reply:
The current implementation makes this function not async safe as it uses a global variable.

Reply:
Done -- changed to pass a a pointer to a struct.